node-UX organize categories

Kolupsy opened the pull request:
I made the default categories deterministic (defined them up front).
Also organized the functions a bit more. I guess the exact order is a bit subjective. Let me know what you think.
Also made it clear from the category UI which nodes have subcategories. This was originally to keep track of where we have subcategories but it might be nice to keep it in as an actual feature.

There are still some functions in the 'Vector' category that I dont know what they do exactly and therefore can not group or move to a different category:
![image](https://user-images.githubusercontent.com/88381426/179395262-cecd1ed0-b7c6-4e2a-94dd-d32b6c49713a.png)

Commit:
organize categories

pragma37 commented:
I find that the subcategory icons actually decrease readability.
I've tried to move them to the right of the label, but it breaks the UI drawing.

The changes look good for the most part, but I think at this point it would be better to just focus on the final nodes rather than keep improving the categories of nodes that will end up marked as internal.

Screen Pixel and Pixel Depth are the pixels (fragment) coordinates.

Pixel World Size is the size of a pixel in world space.

Ray Plane Intersection traces a ray against a plane and returns the ray length at the intersection point.

Kolupsy commented:
> I find that the subcategory icons actually decrease readability.

Alright then lets leave them

>  it would be better to just focus on the final nodes rather than keep improving the categories of nodes that will end up marked as internal.

wdym final nodes? Are those not the final nodes? I know we wanted to use python for some custom nodes but those are mainly the nodes that we are going to bring over from MaltES, right?

> Ray Plane Intersection traces a ray against a plane and returns the ray length at the intersection point.

Oh so its the ray length. Ok I wasnt sure what it was outputting. Seems like it does belong in the vector category.


So the next steps are going to be transferring additional functions over from MaltES and then taking care of the custom node in python?

pragma37 commented:
> wdym final nodes? Are those not the final nodes? 

So, for example, the idea is to use a Subcategory node with a drop-down for Surface, instead of having a Geometry node with multiple outputs like the one from Cycles/EEVEE?

I personally prefer the drop-down option, but I thought that wasn't the plan. 🤔

> Oh so its the ray length. Ok I wasnt sure what it was outputting. Seems like it does belong in the vector category.

Yes, but I think this is one of the nodes that we may want to mark as internal ?

Kolupsy commented:
> So, for example, the idea is to use a Subcategory node with a drop-down for Surface, instead of having a Geometry node with multiple outputs like the one from Cycles/EEVEE?
I personally prefer the drop-down option, but I thought that wasn't the plan. 🤔

hmm maybe this should be discussed a bit more in-depth. Maybe even on a case-by-case basis. I think the 'forces' that are at play here are simplicity, general UX and beginner friendliness. Keeping the Surface node as it is has a lot of simplicity because it doesnt require any extra code to work. UX is probably a bit subjective, I think it is still pretty ergonomic but personally I slighly prefer a node with a wall of sockets (that can be collapsed when unused) over the drop-down. Beginner friendliness is also quite important for Malt right now wouldnt you say? 

How about this: I will try to create the basis for the python custom node next and then make a PR where the nodes are very 'blender-like' (similar to MaltSE). From there we should be able to find the optimal middle-ground between the very procedural system (that we have right now) and a very customized one that maximizes on beginner friendliness. How does that sound?

> Yes, but I think this is one of the nodes that we may want to mark as internal ?

Because it is a very specific function that you will rarily use or because it will be wrapped/replaced with something better?

pragma37 commented:
> Keeping the Surface node as it is has a lot of simplicity because it doesnt require any extra code to work.

Neither does the "wall of sockets". You can just use multiple *out* parameters.

> Beginner friendliness is also quite important for Malt right now wouldnt you say?

Yep, that's the main goal of the redesign.

Something I like about the drop-down approach is that it removes the need for collapsing nodes, and even if you do, it is much more clear what it does (at least if we update the label with the subfunction, which should be trivial to implement).
It's also easier for pipelines and addons to extend the default nodes.
And it allows having even fewer node types (10 outputs for a single node is huge, but 10 subtypes are likely fine).

On the other hand, I have not taken the best UX decisions so far, so I could just be wrong. ¯\_(ツ)_/¯

Maybe I should make a poll?

> Because it is a very specific function that you will rarily use or because it will be wrapped/replaced with something better?

The former. One of the main feedbacks is that having so many nodes is overwhelming,  so I think it would be better to hide the rarely used ones.

Kolupsy commented:
> On the other hand, I have not taken the best UX decisions so far, so I could just be wrong. ¯_(ツ)_/¯

The solution to the issue sure isnt binary so there are no right or wrong decisions in that regard. Clearly blender did also chose a hybrid design by making things like math/vector math operations drop downs and others nodes with socket walls

> Maybe I should make a poll?

Do you think we could get enough results to get a meaningful result? 

> And it allows having even fewer node types (10 outputs for a single node is huge, but 10 subtypes are likely fine).

Keeping the node item count low is probably a good idea but I would bet that most of the overwhelm came from the amount of different categories + technical naming. Even blender is overwhelming at first but after a while you learn all the nodes. 

What about the propsal in my previous message?

pragma37 commented:
I've been testing both approaches:

![Recording 2022-07-18 at 18 29 49](https://user-images.githubusercontent.com/36610999/179559042-bb795905-de01-48cf-a24e-0eed9db84a2d.gif)

> Do you think we could get enough results to get a meaningful result?

Not here, but on twitter I'm sure we would.

> Keeping the node item count low is probably a good idea but I would bet that most of the overwhelm came from the amount of different categories + technical naming.

Yes, I'd bet you're right. But I also think that there are some rare cases where it would be best to just drop the functionality from the default list.

> What about the propsal in my previous message?

I think it would be best to iterate on a few selected nodes first? That way we can figure out a pattern to follow for the rest.

dibli-goost commented:
I'd argue in favor of synchronicity with EEVEE and cycles regarding drop downs and walls of sockets.

Being in line with the UX of those will make coming from those engines much easier to immediately understand, and changing it (to me) leaves little benefit in the way of organization and usability. I personally find myself using a lot of the outputs from the geometry node at once, so being able to do that and not having to add multiple nodes would be preferable.

I think the drop down design works best when the end user will immediately know which output they'll need. If they don't know, then having all those outputs immediately available gives them the ability to more easily swap between each output, trying them out and seeing which one works best for their current situation.

-

Icons in the add menu are also a bit much I think. If we want to add some iconography of sorts there it should be colors, muted colors that don't super catch the eye but instead *direct* the eye. If that makes sense... however I don't know whether or not it's possible *to* assign colors to each of the listings. Worth the thought anyways.

Kolupsy commented:
should we drop this PR then? You already made the main change from this PR in your latest commits...

pragma37 commented:
There are more things in this PR, though?

Regardless, I think we should focus on a new Cycles/EEEVEE/SE-like node API, since that seems to be what most users want, and mark most of the existing functions as internal.

Kolupsy commented:
> There are more things in this PR, though?

Yeah some minor META changes. I can resolve the merge conflict and then you can see if we want to merge this with the `node-ux` branch

Commit:
Merge branch 'node-ux' into node-ux---organize-categories

Commit:
revert subcategory node item icons

Commit:
dont register empty categories

Commit:
better fractal noise

Commit:
rearrange nodes

Kolupsy commented:
there are a few commits here. only the last one `rearrange nodes` has the EEVEE-like node changes. As I said before, its very similar but went a bit further still. I think the Input category would not need any further changes than this.

Here is a rundown of the other commits:

- `revert subcategory node item icons` removes the icons that i added as a test
- `dont register empty categories` I think this might be useful. Especially when we are dealing with the different graphs types. We will likely not need a 'Math' category for the render graph for example
- `better fractal noise` I changed the fractal noise again so that octaves now blender together (renamed to 'detail' and it is a float). I also added a roughness parameter to control the influence of lower/higher frequencies

pragma37 reviewed (changes requested):
I like the new fractal noise features, but let's try to avoid breaking changes.
Just adding a function with the old signature that calls the new one would be enough.

Also, for the node, maybe we should split the vec4 coord into vec3 coord and float seed/time?

pragma37 commented on Malt/Shaders/Common/Transform.glsl:
```diff
@@ -49,19 +49,19 @@ vec3 transform_normal(mat4 matrix, vec3 normal)
 }
 
-/* META @meta: category=Input; subcategory=Camera Info; */
+/* META @meta: category=Input; subcategory=Camera Info; internal=true;*/
```
For files where everything or almost everything is internal, we should use a global meta property.

Kolupsy replied:
got it. Will change.

pragma37 commented on Malt/Pipelines/NPR_Pipeline/Shaders/NPR_Pipeline/NPR_Mesh.glsl:
```diff
@@ -152,5 +152,15 @@ bool is_main_pass()
 }
+/*META @meta: category=Input; */
+void pass_info(
+    out bool main_pass,
```
For bool properties I would like to use the *Is* naming, so *Shadow Pass* -> *Is Shadow Pass*.

Kolupsy replied:
having the out variable have the same name as a function doesnt work apparently. Is there a way around it? or would we just make META labels for those outputs?

pragma37 replied:
You can name the out variable Is_Main_Pass, or simply use a label.

Kolupsy replied:
How do you add a label to an output? Can you even do that? Because generall only the inputs/parameters can have meta data right? Because of that I went with the first option

pragma37 replied:
Any function parameter can have metadata.
This should work:
````
@main_pass: label=Is Main Pass;
````

pragma37 commented on Malt/Shaders/Node Utils/Input.glsl:
```diff
@@ -0,0 +1,97 @@
+    int uv_index,
+    out vec3 position,
+    out vec3 normal,
+    out vec3 tangent,
```
I would avoid having duplicated properties.

Tangent/Bitangent belong to its own node IMO, since it requires extra parameters that this node ignores.

Incoming is the same as -view_direction(), so the Camera Data View Vector should be enough,

Facing should be in the Fresnel/Layer Weight node.

Kolupsy replied:
thinking about it I have to agree with the tangent/bitangent. Lets change that.

incoming is being used a lot, I would like to have it on the geometry node. But I can remove the `get_incoming` and use the already available function.

do we have a fresnel function? If Facing would be a single node with a single output I would rather have it on the geometry node.

pragma37 replied:
Ok, we can leave the incoming output.

We don't have a fresnel function, but I think we should.

Kolupsy replied:
I only know the one blender uses. Its probably general enough to use (because of the licensing thing)?

pragma37 replied:
Do you mean the actual implementation?
We have the ones from here:
https://github.com/bnpr/Malt/blob/Development/Malt/Shaders/Shading/BRDF.glsl#L159
Ported from here:
http://graphicrants.blogspot.com/2013/08/specular-brdf-reference.html

But I'm not sure how they map to the IOR, tbh.

pragma37 commented on Malt/Shaders/Node Utils/Input.glsl:
```diff
@@ -0,0 +1,97 @@
+        original_id = IO_ID;
+    }
+
+int _get_fps(){ return int(round(FRAME / TIME));}
```
Would break at frame 0.

Kolupsy replied:
true. Is ther any other way to get the frame rate? if not then lets just scap it. its not important enough to let this inconsistency pass imo

pragma37 replied:
We would have to retrieve the info from Blender. I would simply drop it for now.

Kolupsy replied:
sounds good 👍

Commit:
node categories cleanup

Kolupsy commented:
I think I took all your requests into account. I did some additional META cleanup especially for labels and min/max values.

pragma37 committed:
fix typo

pragma37 committed:
Remove Slider from AO bias

bias should be kept as possible

pragma37 commented:
Looks good.
Can we move this to the main pull request, so we can iterate on it at the same time?

Kolupsy commented:
> Looks good.
> 
> Can we move this to the main pull request, so we can iterate on it at the same time?

Yes please 🙂

Kolupsy commented:
oh thanks for merging that already
